### DISCUSSION TITLE
cannot get NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE work with smart pointer

### DISCUSSION THREAD

**Discussion by obi0ne** [nlohmann/json#2964] [Q&A]
Hi,

Need help, I cannot get `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE` to work with smart pointers.
See this small sample code to demonstrate the problem:

```
#include <iostream>
#include <memory>
#include "json.hpp"

using json = nlohmann::json;

namespace nlohmann {
	template <typename T>
	struct adl_serializer<std::shared_ptr<T>> {

		static void to_json(json& j, const std::shared_ptr<T>& opt) {
			if (opt.get()) {
				j = *opt;
			}
			else {
				j = nullptr;
			}
		}

		static void from_json(json& j, std::shared_ptr<T>& ptr) {
			if (j.is_null()) {
				ptr.reset();
			}
			else {
				ptr = std::make_shared<T>(j.get<T>());
			}
		}
	};
}

typedef struct tag_Foo
{
    unsigned int a;
    unsigned int b;
} Foo;

typedef struct tagBar
{
    std::shared_ptr<Foo> myFoo;
} Bar;



NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Foo
	, a
	, b
);


NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Bar
	, myFoo
);



int main()
{
	std::shared_ptr<int> a = std::make_shared<int>(1);
	std::shared_ptr<int> b;
	json j = { a, b };

	Bar bar;

	json j2 = bar;
}
```

The following code produce following compilition error on VS2019 version 16.11.0):

> 1>DebugJsonSerialization.cpp
> 1>C:\Users\xxxxx\source\repos\DebugJsonSerialization\DebugJsonSerialization\DebugJsonSerialization.cpp(53,1): error C2672: 'nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to': no matching overloaded function found
> 1>C:\Users\xxxxx\source\repos\DebugJsonSerialization\DebugJsonSerialization\DebugJsonSerialization.cpp(53,1): error C2783: 'ValueType &nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to(ValueType &) noexcept(<expr>) const': could not deduce template argument for '__formal'
> 1>C:\Users\xxxxx\source\repos\DebugJsonSerialization\DebugJsonSerialization\json.hpp(20658): message : see declaration of 'nlohmann::basic_json<std::map,std::vector,std::string,bool,int64_t,uint64_t,double,std::allocator,nlohmann::adl_serializer,std::vector<uint8_t,std::allocator<uint8_t>>>::get_to'

What I am doing wrong, can anyone help ?

**Comment by nlohmann** [comment]
The problem is the signature of your adl_serializer<std::shared_ptr<T>>::from_json: you wrote `static void from_json(json& j, std::shared_ptr<T>& ptr)`, taking `j` by non-const reference. The library's `has_from_json` trait (include/nlohmann/detail/meta/type_traits.hpp) only detects a `from_json` overload whose first parameter is exactly `const BasicJsonType&`; with a non-const `json&` parameter your specialization is invisible to that trait, so `get_to`/`get<std::shared_ptr<Foo>>()` has no valid overload to call, producing the C2672/C2783 errors. The fix is simply to make the parameter const: `static void from_json(const json& j, std::shared_ptr<T>& ptr)`. This matches every adl_serializer example in docs/mkdocs/docs/features/arbitrary_types.md (e.g. the boost::optional<T> example), which all use `const json&`/`const BasicJsonType&`.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
